apps/src/lib/ui/ChangeEmailModal.jsx

```diff
@@ -0,0 +1,341 @@
+A less clumsy way to use Rails UJS?
+
+ */
+
+const STATE_INITIAL = 'initial';
+const STATE_SAVING = 'saving';
+const STATE_UNKNOWN_ERROR = 'unknown-error';
+
+export default class ChangeEmailModal extends React.Component {
+  static propTypes = {
+    isOpen: PropTypes.bool.isRequired,
+    handleSubmit: PropTypes.func.isRequired,
+    handleCancel: PropTypes.func.isRequired,
+    railsForm: PropTypes.object,
+    userAge: PropTypes.number.isRequired,
+    currentHashedEmail: PropTypes.string,
+  };
+
+  state = {
+    saveState: STATE_INITIAL,
+    newEmail: '',
+    newEmailServerError: undefined,
+    currentPassword: '',
+    currentPasswordServerError: undefined,
+    emailOptIn: '',
+  };
+
+  //
+  // Note: This dialog submits account changes to dashboard using a
+  // hidden Rails-generated form.  It expects that form to already exist in the
+  // DOM when the component mounts and to have a particular data attribute.
+  //
+  // When the user clicks the "update" button, this dialog loads the relevant
+  // information into the hidden Rails form and calls submit(). Rails injects
+  // all the JavaScript needed for the form to submit via AJAX with all the
+  // appropriate validation tokens, etc.  The dialog subscribes to events
+  // emitted by the Rails helper JavaScript to detect success or errors.
+  //
+  // If the dialog can't find the Rails form anywhere it will emit a warning
+  // and be unable to submit anything (useful for tests and storybook).
+  //
+  // Read more:
+  // http://guides.rubyonrails.org/working_with_javascript_in_rails.html#rails-ujs-event-handlers
+  // https://github.com/rails/jquery-ujs
+  //
+
+  componentDidMount() {
+    this._form = $(this.props.railsForm);
+    if (this._form) {
+      this._form.on('ajax:success', this.onSubmitSuccess);
+      this._form.on('ajax:error', this.onSubmitFailure);
+    } else {
+      console.warn('The ChangeEmailModal component did not find the required ' +
+        'Rails UJS form, and will be unable to submit changes.');
+    }
+    this.newEmailInput.focus();
+  }
+
+  componentWillUnmount() {
+    if (this._form) {
+      this._form.off('ajax:success', this.onSubmitSuccess);
+      this._form.off('ajax:error', this.onSubmitFailure);
+      delete this._form;
+    }
+  }
+
+  save = () => {
+    if (!this._form) {
+      console.error('The ChangeEmailModal component did not find the required ' +
+        'Rails UJS form, and was unable to submit changes.');
+      return;
+    }
+
+    this.setState({saveState: STATE_SAVING}, () => {
+      this._form.find('#change-email-modal_user_email').val(this.props.userAge < 13 ? '' : this.state.newEmail);
+      this._form.find('#change-email-modal_user_hashed_email').val(hashEmail(this.state.newEmail));
+      this._form.find('#change-email-modal_user_current_password').val(this.state.currentPassword);
+      // this._form.find('#user_email_opt_in').val(this.state.emailOptIn);
```

Comment:
why commented out? Is this a future field?

Reply:
Yes.  The whole reason we're converting this to a dialog is so that we can explicitly ask teachers to re-opt-in to marketing email when they change their email address.  I started building this out, then realized just switching to a dialog for this flow was complicated enough, so I'll be adding the opt-in as a follow-on PR.  I wasn't sure it was worth extracting the work I'd done so far though.